tools/codegen/dest/register_dispatch_key.py

```diff
@@ -84,15 +84,34 @@ def gen_resize_out_helper(backend_index: BackendIndex) -> List[str]:
     } else if (options.memory_format_opt().has_value()) {
       out.unsafeGetTensorImpl()->empty_tensor_restride(*options.memory_format_opt());
     }
+  } else {
+    at::assert_no_internal_overlap(out);
   }
 }
 """]
 
+def gen_check_inplace_helper(backend_index: BackendIndex) -> List[str]:
+    return ["""
+void check_inplace(const Tensor &self, IntArrayRef sizes, const TensorOptions &options) {
+  TORCH_CHECK(options.dtype() == self.dtype(),
+      "Bad in-place call: ",
+      "input tensor dtype ", self.dtype(), " and output tensor dtype ", options.dtype(), " should match");
+  TORCH_CHECK(options.device() == self.device(),
+      "Bad in-place call: ",
+      "input tensor device ", self.device(), " and output tensor device ", options.device(), " should match");
+  TORCH_CHECK(sizes == self.sizes(),
+      "Bad in-place call: ",
+      "input tensor size ", self.sizes(), " and output tensor size ", sizes, " should match");
+  at::assert_no_internal_overlap(self);
+}
```

Comment:
There's a subtlety here which is probably worth adding comments for, which is that some of the tests that you see here are done separately for TensorIterator kernels, inside TensorIterator. So while this is required for things like addmm (which don't do TensorIterator), it is redundant for other operators (like add) and are a slight regression in instruction count. I don't think this PR should get blocked because of this, especially because ufunc codegen is coming along well and should make it easier to fix this problem, it's just something to be aware of.